crates/ruff_linter/src/rules/pycodestyle/rules/blank_lines.rs

```diff
@@ -0,0 +1,925 @@
+                        count: _blank_lines,
+                        range: blank_lines_range,
+                    } => {
+                        diagnostic.set_fix(Fix::safe_edit(Edit::range_replacement(
+                            stylist
+                                .line_ending()
+                                .repeat((BLANK_LINES_TOP_LEVEL) as usize),
+                            blank_lines_range,
+                        )));
+                    }
+                    BlankLines::Zero => {
+                        diagnostic.set_fix(Fix::safe_edit(Edit::insertion(
+                            stylist
+                                .line_ending()
+                                .repeat((BLANK_LINES_TOP_LEVEL) as usize),
+                            locator.line_start(self.last_non_comment_line_end),
+                        )));
+                    }
+                }
+
+                diagnostics.push(diagnostic);
+            }
+
+            if line.blank_lines > BLANK_LINES_TOP_LEVEL
+                || (line.indent_length > 0 && line.blank_lines > BLANK_LINES_METHOD_LEVEL)
+            {
+                // E303
+                let mut diagnostic = Diagnostic::new(
+                    TooManyBlankLines {
+                        actual_blank_lines: line.blank_lines.count(),
+                    },
+                    line.first_token_range,
+                );
+
+                if let BlankLines::Many {
+                    count: _,
+                    range: blank_lines_range,
+                } = line.blank_lines
+                {
+                    if line.indent_length > 0 {
+                        diagnostic.set_fix(Fix::safe_edit(Edit::range_replacement(
+                            stylist
+                                .line_ending()
+                                .repeat((BLANK_LINES_METHOD_LEVEL) as usize),
+                            blank_lines_range,
+                        )));
+                    } else {
+                        diagnostic.set_fix(Fix::safe_edit(Edit::range_replacement(
+                            stylist
+                                .line_ending()
+                                .repeat((BLANK_LINES_TOP_LEVEL) as usize),
+                            blank_lines_range,
+                        )));
+                    };
+                }
+
+                diagnostics.push(diagnostic);
+            }
+
+            if matches!(self.follows, Follows::Decorator)
+                && !line.is_comment_only
+                && line.preceding_blank_lines > 0
+            {
+                // E304
+                let mut diagnostic =
+                    Diagnostic::new(BlankLineAfterDecorator, line.first_token_range);
+
+                // Get all the lines between the last decorator line (included) and the current line (included).
+                // Then remove all blank lines.
+                let trivia_range = locator.lines_range(TextRange::new(
```

Comment:
I don't think using `stylist.line_ending().text_len()` is safe in case the file uses mixed newlines because the first newline could be `\n` but the newline after the decorator might be `\r\n` (the problem we had with other blank line replacement bugs as well)

What are we trying to do here? 

Shouldn't `self.last_non_comment_line_end` already give you the end position of the decorator line? 

What seems to be the easiest fix is to use 

```rust
               diagnostic.set_fix(Fix::safe_edit(Edit::range_deletion(
                    without_blank_lines.to_string(),
                    TextRange::new(
                        self.last_non_comment_line_end,
                        line.first_token_range.start(),
                    ),
                )));
```

but it doesn't handle the case where you have a comment in the range between (which the old implementation handles). However, the old implementation only handles `\n` but not `\r` or `\r\n`. 
We would need to adopt your fix to handle any possible empty line sequence and replace it with `stylist.newline` (and not `\n`)

To help future readers understand this code better, it might help to add a example code snipped to the outer if branch: like

```rust
// Removes unnecessary empty lines between the decorator and a function or class definition
// ```python
// @decorator
// 
// def test(): pass
// ```
```

Reply:
Okay. I pushed a version that addresses the line ending issue. A few inline documentation blocks examples with examples might still be useful